pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -0,0 +1,371 @@
+   * taken till the measure is being reported.
+  *
+  * @param currentDelay original sample delay to which we will add the age of the measure.
+   */
+  private long getAgedDelay(DelayMeasure currentDelay) {
+    if (currentDelay == null) {
+      return 0; // return 0 when not initialized
+    }
+    // Add age of measure to the reported value
+    long measureAgeInMs = _clock.millis() - currentDelay._sampleTime;
+    // Correct to zero for any time shifts due to NTP or time reset.
+    measureAgeInMs = Math.max(measureAgeInMs, 0);
+    return currentDelay._delayMilliseconds + measureAgeInMs;
+  }
+
+  /*
+   * Helper function to be called when we should stop tracking a given partition. Removes the partition from
+   * all our maps.
+   *
+   * @param partitionGroupId partition ID which we should stop tracking.
+   */
+  private void removePartitionId(int partitionGroupId) {
+    _partitionToDelaySampleMap.remove(partitionGroupId);
+    // If we are removing a partition we should stop reading its ideal state.
+    _partitionsMarkedForVerification.remove(partitionGroupId);
+    if (_enablePerPartitionMetric) {
+      _serverMetrics.removeTableGauge(getPerPartitionMetricName(partitionGroupId),
+          ServerGauge.TABLE_INGESTION_DELAY_MS);
+    }
+  }
+
+  /*
+   * Helper function to generate a per partition metric name.
+   *
+   * @param partitionGroupId the partition group id to be appended to the table name so we
+   *        can differentiate between metrics for various partitions.
+   *
+   * @return a metric name with the following structure: _metricName + partitionGroupId
+   */
+  private String getPerPartitionMetricName(int partitionGroupId) {
+    return _metricName + partitionGroupId;
+  }
+
+  /*
+   * Helper functions that creates a list of all the partitions that are marked for verification and whose
+   * timeouts are expired. This helps us optimize checks of the ideal state.
+   */
+  private ArrayList<Integer> getPartitionsToBeVerified() {
+    ArrayList<Integer> partitionsToVerify = new ArrayList<>();
+    for (int partitionGroupId : _partitionsMarkedForVerification.keySet()) {
+      long markTime = _partitionsMarkedForVerification.get(partitionGroupId);
+      long timeMarked = _clock.millis() - markTime;
+      if (timeMarked > PARTITION_TIMEOUT_MS) {
+        // Partition must be verified
+        partitionsToVerify.add(partitionGroupId);
+      }
+    }
+    return partitionsToVerify;
+  }
+
+  // Custom Constructor
+  public IngestionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
+      RealtimeTableDataManager realtimeTableDataManager, int timerThreadTickIntervalMs, String metricNamePrefix,
+      boolean enableAggregateMetric, boolean enablePerPartitionMetric)
+      throws RuntimeException {
+    _logger = LoggerFactory.getLogger(getClass().getSimpleName());
+    _serverMetrics = serverMetrics;
+    _tableNameWithType = tableNameWithType;
+    _metricName = metricNamePrefix + tableNameWithType;
```

Comment:
Iiuc, `ServerMetrics` already has a metric prefix that can be configured. Each partition level metric includes the `partitionGroupId`.  how will this `metricNamePrefix` will be used? Do we really need it?

Reply:
this is a bit forwarding looking to extend this work to create another metric that report end to end ingestion delays as our offline discussion

Reply:
I'd suggest keeping things simple especially for the first version. We can extend it in the future after getting the first version stable. We may consider removing `metricNamePrefix`, `enableAggregateMetric` and `enablePerPartitionMetric` and always emit partition level metric

Reply:
Done.

Reply:
significantly simplified all this, see 3ce7b783c5dc88073106f400f00783ba4ab71281